libPhoneNumberGeocoding/NBPhoneNumberOfflineGeocoder.m

```diff
@@ -0,0 +1,150 @@
+  }
+  return self;
+}
+
+- (nullable NSString *)countryNameForNumber:(NBPhoneNumber *)number
+                           withLanguageCode:(NSString *)languageCode {
+  NSArray *regionCodes = [_phoneNumberUtil getRegionCodesForCountryCode:number.countryCode];
+  if ([regionCodes count] == 1) {
+    return [self regionDisplayName:regionCodes[0] withLanguageCode:languageCode];
+  } else {
+    NSString *regionWhereNumberIsValid = @"ZZ";
+    for (NSString *regionCode in regionCodes) {
+      if ([_phoneNumberUtil isValidNumberForRegion:number regionCode:regionCode]) {
+        if (![regionWhereNumberIsValid isEqualToString:@"ZZ"]) {
+          return nil;
+        }
+        regionWhereNumberIsValid = regionCode;
+      }
+    }
+
+    return [self regionDisplayName:regionWhereNumberIsValid withLanguageCode:languageCode];
+  }
+}
+
+- (nullable NSString *)regionDisplayName:(NSString *)regionCode
+                        withLanguageCode:(NSString *)languageCode {
+  if (regionCode == nil || [regionCode isEqualToString:@"ZZ"] ||
+      [regionCode isEqual:NB_REGION_CODE_FOR_NON_GEO_ENTITY]) {
+    return nil;
+  } else {
+    return [[NSLocale localeWithLocaleIdentifier:languageCode] displayNameForKey:NSLocaleCountryCode
+                                                                           value:regionCode];
+  }
+}
+
+- (nullable NSString *)descriptionForValidNumber:(NBPhoneNumber *)phoneNumber
+                                withLanguageCode:(NSString *)languageCode {
+  // If the NSCache doesn't contain a key equivalent to languageCode, create a new
+  // NBGeocoderMetadataHelper object with a language set equal to languageCode and
+  // default country code to United States / Canada
+  if ([_metadataHelpers objectForKey:languageCode] == nil) {
+    [_metadataHelpers setObject:[[NBGeocoderMetadataHelper alloc] initWithCountryCode:@1
+                                                                         withLanguage:languageCode]
+                         forKey:languageCode];
+  }
+  NSString *ans = [[_metadataHelpers objectForKey:languageCode] searchPhoneNumber:phoneNumber];
+  if (ans == nil) {
+    return [self countryNameForNumber:phoneNumber withLanguageCode:languageCode];
+  } else {
+    return ans;
+  }
+}
+
+- (nullable NSString *)descriptionForValidNumber:(NBPhoneNumber *)phoneNumber
+                                withLanguageCode:(NSString *)languageCode
+                                  withUserRegion:(NSString *)userRegion {
+  NSString *regionCode = [_phoneNumberUtil getRegionCodeForNumber:phoneNumber];
+  NSLog(@"The region code from phone util was: %@, compared to parameter: %@", regionCode,
+        userRegion);
+  if ([userRegion isEqualToString:regionCode]) {
+    return [self descriptionForValidNumber:phoneNumber withLanguageCode:languageCode];
+  }
+  return [self regionDisplayName:regionCode withLanguageCode:languageCode];
+}
+
+- (nullable NSString *)descriptionForNumber:(NBPhoneNumber *)phoneNumber
```

Comment:
rusty on obj-c, but can this just call into 
```
- (nullable NSString *)descriptionForNumber:(NBPhoneNumber *)phoneNumber
                           withLanguageCode:(NSString *)languageCode
                             withUserRegion:(NSString *)userRegion
```

with a nil userRegion?

Reply:
yeah it should be able to

Reply:
if userRegion is nil, should i react by just returning: return `[self regionDisplayName:regionCode withLanguageCode:languageCode];` ? @aalexli @paween

Reply:
Actually I think the problem is that regionCode is non-null in the API so if we pass in nil the userRegion then it breaks the API contract. I think it's fine to leave it like this for now and we can think of a way to optimize that in the future.

Reply:
I would think only this method changes. 

Just pass nil as userRegion and don't think about it again :)

the final method that checks for nil (which it should be doing for the current one) will handle it.